holoviews/element/util.py

```diff
@@ -30,3 +44,98 @@ def compute_edges(edges):
         raise ValueError('Centered bins have to be of equal width.')
     edges -= width/2.
     return np.concatenate([edges, [edges[-1]+width]])
+
+
+def reduce_fn(x):
+    """
+    Aggregation function to get the first non-zero value.
+    """
+    values = x.values if pd and isinstance(x, pd.Series) else x
+    for v in values:
+        if not is_nan(v):
+            return v
+    return np.NaN
+
+
+class categorical_aggregate2d(ElementOperation):
+    """
+    Generates a gridded Dataset of 2D aggregate arrays indexed by the
+    first two dimensions of the passed Element, turning all remaining
+    dimensions into value dimensions. The key dimensions of the
+    gridded array are treated as categorical indices. Useful for data
+    indexed by two independent categorical variables such as a table
+    of population values indexed by country and year. Data that is
+    indexed by continuous dimensions should be binned before
+    aggregation. The aggregation will retain the global sorting order
+    of both dimensions.
+
+    >> table = Table([('USA', 2000, 282.2), ('UK', 2005, 58.89)],
+                     kdims=['Country', 'Year'], vdims=['Population'])
+    >> categorical_aggregate2d(table)
+    Dataset({'Country': ['USA', 'UK'], 'Year': [2000, 2005],
+             'Population': [[ 282.2 , np.NaN], [np.NaN,   58.89]]},
+            kdims=['Country', 'Year'], vdims=['Population'])
+    """
+
+    datatype = param.List(['xarray', 'grid'] if xr else ['grid'], doc="""
+        The grid interface types to use when constructing the gridded Dataset.""")
+
+    def _process(self, obj, key=None):
+        """
+        Generates a categorical 2D aggregate by inserting NaNs at all
+        cross-product locations that do not already have a value assigned.
+        Returns a 2D gridded Dataset object.
+        """
```

Comment:
Quite a long method...if you see chunks that could be split up into helper methods, that might be sensible. Up to you though!